File: packages/core/src/single-select.ts
/**
 * Single-select behavior machine, shared by RadioGroup, SegmentedControl, and
 * TriStateSwitch (contracts: radio-group.md, segmented-control.md,
 * tri-state-switch.md, "Behavior Machine" sections).
 *
 * One implicit state; the selected value lives in context. Selecting the
 * current value is inert — native radio inputs cannot re-fire an unchanged
 * selection, so the guard is behavior-invariant for UI paths.
 */

export interface SelectOption {
  value: string;
  disabled?: boolean;
}

export interface SingleSelectContext {
  value: string | null;
  options: SelectOption[];
  disabled: boolean;
}

export type SingleSelectEvent =
  | { type: "SELECT"; value: string }
  | { type: "SET_VALUE"; value: string | null };

export type SingleSelectEffect = { type: "emitValueChange"; value: string };

export interface SingleSelectResult {
  context: SingleSelectContext;
  effects: SingleSelectEffect[];
}

export function singleSelectTransition(
  context: SingleSelectContext,
  event: SingleSelectEvent,
): SingleSelectResult {
  switch (event.type) {
    case "SELECT": {
      const option = context.options.find((candidate) => candidate.value === event.value);

      if (context.disabled || !option || option.disabled || context.value === event.value) {
        return { context, effects: [] };
      }

      return {
        context: { ...context, value: event.value },
        effects: [{ type: "emitValueChange", value: event.value }],
      };
    }
    case "SET_VALUE": {
      return { context: { ...context, value: event.value }, effects: [] };
    }
  }
}

File: packages/contracts/headless/src/single_select.rs
//! Single-select machine (RadioGroup, SegmentedControl, TriStateSwitch).
//! Mirror of core `single-select.ts`.

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SelectOption {
    pub value: String,
    pub disabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SingleSelectContext {
    pub value: Option<String>,
    pub options: Vec<SelectOption>,
    pub disabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SingleSelectEvent {
    Select { value: String },
    SetValue { value: Option<String> },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SingleSelectEffect {
    EmitValueChange { value: String },
}

pub fn single_select_transition(
    context: SingleSelectContext,
    event: SingleSelectEvent,
) -> (SingleSelectContext, Vec<SingleSelectEffect>) {
    match event {
        SingleSelectEvent::Select { value } => {
            let option = context
                .options
                .iter()
                .find(|candidate| candidate.value == value);
            let same = context.value.as_deref() == Some(value.as_str());

            match option {
                Some(option) if !context.disabled && !option.disabled && !same => {
                    let effects = vec![SingleSelectEffect::EmitValueChange {
                        value: value.clone(),
                    }];

                    (
                        SingleSelectContext {
                            value: Some(value),
                            ..context
                        },
                        effects,
                    )
                }
                _ => (context, vec![]),
            }
        }
        SingleSelectEvent::SetValue { value } => (SingleSelectContext { value, ..context }, vec![]),
    }
}
